homeassistant/components/demo/number.py

```diff
@@ -0,0 +1,83 @@
+"""Demo platform that offers a fake Number entity."""
+from homeassistant.components.number import NumberEntity
+from homeassistant.const import DEVICE_DEFAULT_NAME
+
+from . import DOMAIN
+
+
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
+    """Set up the demo Number entity."""
+    async_add_entities(
+        [
+            DemoNumber(
+                "volume1",
+                "volume",
+                42.0,
+                "mdi:volume-high",
+                False,
+            ),
+        ]
+    )
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the Demo config entry."""
+    await async_setup_platform(hass, {}, async_add_entities)
+
+
+class DemoNumber(NumberEntity):
+    """Representation of a demo Number entity."""
+
+    def __init__(self, unique_id, name, state, icon, assumed):
+        """Initialize the Demo Number entity."""
+        self._unique_id = unique_id
+        self._name = name or DEVICE_DEFAULT_NAME
+        self._state = state
+        self._icon = icon
+        self._assumed = assumed
+
+    @property
+    def device_info(self):
+        """Return device info."""
+        return {
+            "identifiers": {
+                # Serial numbers are unique identifiers within a specific domain
+                (DOMAIN, self.unique_id)
+            },
+            "name": self.name,
+        }
+
+    @property
+    def unique_id(self):
+        """Return the unique id."""
+        return self._unique_id
+
+    @property
+    def should_poll(self):
+        """No polling needed for a demo Number entity."""
+        return False
+
+    @property
+    def name(self):
+        """Return the name of the device if any."""
+        return self._name
+
+    @property
+    def icon(self):
+        """Return the icon to use for device if any."""
+        return self._icon
+
+    @property
+    def assumed_state(self):
+        """Return if the state is based on assumptions."""
+        return self._assumed
+
+    @property
+    def state(self):
+        """Return the current value."""
+        return self._state
+
+    async def async_set_value(self, value):
+        """Update the current value."""
+        self._state = value
```

Comment:
Maybe it doesn't matter because this is just for a demo, but you may want to check that `value` is within the range.

Reply:
Agree. It's not just a demo, it's more like a reference that platforms will base their design on.
Fixed.